### DISCUSSION TITLE
next/image component not working within a dynamic import in /app/ and Next 13.

### DISCUSSION THREAD

**Discussion by ianhempsell** [vercel/next.js#46653] [Help] [answered]
### Summary

I am dynamically loading my components via next/dynamic and I have a simple TextImage component that I'm having an issue with.

If I render {image} within a html img tag like this: `<img src={image} alt='' />` the image renders on the page wihout issue.

But if I try to use the <Image /> component like this:

```
<Image
    src={image}
    alt=''
    width={640}
    height={360}
/>
```

I get the following error:

`Error: Cannot read properties of null (reading 'useContext')`

I am using WebPack not Turbo.

Has anybody had this issue with the latest release of Next 13.2.3?

### Additional information

```js
This is the error within the console:
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
error - TypeError: Cannot read properties of null (reading 'useContext')
    at stringify (<anonymous>)
error - TypeError: Cannot read properties of null (reading 'useContext')
    at stringify (<anonymous>)
error - TypeError: Cannot read properties of null (reading 'useContext')
    at stringify (<anonymous>)
digest: "2317118846"
error - TypeError: Cannot read properties of null (reading 'useContext')
    at stringify (<anonymous>)
digest: "2317118846"
```

**Comment by ciruz** [comment]
Hey it's me again. Do you have a little bit more code? There must be something else wrong.

But I will give it a try tomorrow, maybe it's a bug, sadly I'm on mobile now.

**Reply by ianhempsell** [reply]
Hey T, I'll try to clean up my code and send it to you. Thanks for reaching out.

**Reply by ianhempsell** [reply]
Hey T, I've created a repo with the issue: https://github.com/ianhempsell/dynamic-blocks-next13.git
If you have chance to look that would be awesome.
Thanks, Ian.

**Reply by ciruz** [reply]
I get the same error like you.

When I try the first approach I recommended you, it works with `next/image`.

Something like this works fine:
 
```tsx
import dynamic from 'next/dynamic';

const TextComponent = dynamic(() => import("./Text"));
const ImageComponent = dynamic(() => import("./TextImage"));
const Wysiwyg = dynamic(() => import("./Wysiwyg"));

const Components = {
  Text: TextComponent ,
  TextImage: ImageComponent ,
  Wysiwyg: Wysiwyg
};

export default function Page() {
    const acfFc = {
       // ...
    };

    return (
        <main className='container mx-auto'>
            {/* ACFFC */}
            {acfFc.layout.blocks.map((item) => {
                const { block_id, block_name, fields } = item;

                const Component = Components[block_name];

                return (
                    <Component
                        key={block_id}
                        blockId={block_id}
                        fields={fields}
                    />);
            })}
        </main>
    );
```

I also found out that:

> dynamic() can't be used inside of React rendering as it needs to be marked in the top level of the module for preloading to work, similar to React.lazy. https://nextjs.org/docs/advanced-features/dynamic-import#example

So for me it's now more surprising that variant 2 worked at all.

**Reply by ianhempsell** [reply]
Thanks T.
I guess I cannot 'dynamically' load/include only the component I need on a pre page basis, I'll have to manually include all available componests so if they are within the 'layout' object, then they will be available.
Shame, I thought JS would have been able to handle this.
Thank for your time. Ian.

**Comment by ianhempsell** [comment, marked as answer]
@ciruz Thanks for your help with this.
I have refactored the basic concept further and just thought I'd share with you:

```Typescript
import dynamic from 'next/dynamic';

const data = ['Text', 'TextImage', 'Text', 'TextImage', 'Text'];

interface Components {
    [key: string]: any;
}

const components: Components = data.reduce((components, component) => {
    return {
        ...components,
        [component]: dynamic(() => import(`./${component}`)),
    };
}, {});

export default function Page() {
    return (
        <main className='container mx-auto'>
            {data.map((component, index) => {
                const Component = components[component];
                const key = `${component}-${index}`;

                return <Component key={key} />;
            })}
        </main>
    );
}
```

**Reply by ciruz** [reply]
nice
